nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java

```diff
@@ -202,40 +219,73 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
 
         try {
             final String transform = readTransform(transformProperty, original);
             final Expression jsltExpression = transformCache.get(transform, currString -> Parser.compileString(transform));
+            final boolean prettyPrint = context.getProperty(PRETTY_PRINT).asBoolean();
 
-            final JsonNode transformedJson = jsltExpression.apply(jsonNode);
-            final ObjectWriter writer = context.getProperty(PRETTY_PRINT).asBoolean() ? jsonObjectMapper.writerWithDefaultPrettyPrinter() : jsonObjectMapper.writer();
-            final Object outputObject;
-            if (transformedJson == null || transformedJson.isNull()) {
-                getLogger().warn("JSLT Transform resulted in no data {}", original);
-                outputObject = null;
-            } else {
-                outputObject = transformedJson;
-            }
-            FlowFile transformed = session.write(original, out -> {
-                if (outputObject != null) {
-                    writer.writeValue(out, outputObject);
+            transformed = session.write(original, (inputStream, outputStream) -> {
+                boolean topLevelArray = false;
+                JsonParser jsonParser;
+                JsonNode firstJsonNode;
+                if (APPLY_TRANSFORM_TO_EACH_OBJECT.equals(transformStrategy)) {
+                    jsonParser = jsonFactory.createParser(inputStream);
+                    jsonParser.setCodec(jsonObjectMapper);
+
+                    JsonToken token = jsonParser.nextToken();
+                    if (token == JsonToken.START_ARRAY) {
+                        token = jsonParser.nextToken(); // advance to START_OBJECT token
+                        topLevelArray = true;
+                    }
+                    if (token == JsonToken.START_OBJECT) { // could be END_ARRAY also
+                        firstJsonNode = jsonParser.readValueAsTree();
+                    } else {
+                        firstJsonNode = null;
+                    }
+                } else {
+                    firstJsonNode = readJson(inputStream);
+                    jsonParser = null; // This will not be used when applying the transform to the entire FlowFile
+                }
+
+                final ObjectWriter writer = prettyPrint ? jsonObjectMapper.writerWithDefaultPrettyPrinter() : jsonObjectMapper.writer();
+                final JsonGenerator jsonGenerator = writer.createGenerator(outputStream);
+
+                Object outputObject;
+                JsonNode nextNode;
+
+                if (topLevelArray) {
+                    jsonGenerator.writeStartArray();
                 }
+                nextNode = firstJsonNode;
+                do {
+                    final JsonNode transformedJson = jsltExpression.apply(nextNode);
+                    if (transformedJson == null || transformedJson.isNull()) {
+                        getLogger().warn("JSLT Transform resulted in no data {}", original);
```

Comment:
This could be repetitive with a large number of records, perhaps set a flag an log once after the loop?